Rewrote a lot of the usage documentation

ewels opened the pull request:
A lot of the usage documentation on the website was getting old. Some was plain wrong, a lot was duplicated and much of it was tricky to navigate and not locally ordered.

Here I have had a stab at cleaning up the usage documentation. I deleted a lot of stuff, wrote a few new things and generally tried to make things a bit more logical.

Oversight welcome! I changed a lot of stuff here 😆 Probably easier to read the documentation as a whole on my fork: https://github.com/ewels/nf-co.re/tree/usage-docs-revamp/markdown/usage

ewels commented:
Hmm, perhaps I was a little hasty in removing the Nextflow tutorial. My logic was that we should instead redirect to the central Nextflow website for this, as it's not really nf-core material.

However, it seems that quite a lot of people are googling it. These are the page views for the whole website for the past month:
![image](https://user-images.githubusercontent.com/465550/73673547-155f2700-46af-11ea-863a-5195f9fbf20b.png)

Note that nearly all of the Nextflow tutorial hits are coming directly from google:
![image](https://user-images.githubusercontent.com/465550/73673896-bcdc5980-46af-11ea-9b79-d70441cf095f.png)

So it might just be a case of us being more googlable than the alternatives.

I guess we have a few options:

1. We just delete it - in the PR currently visiting that URL will redirect to the _nf-core_ tutorial instead
2. We keep the page there as it is
3. We keep a skeleton page with info of where to find other tutorials

Thoughts? Votes?

drpatelh commented on markdown/usage/installation.md:
````diff
@@ -31,125 +18,52 @@ java -version
-```
+An analysis pipeline chains the execution of multiple tools together.
+Historically, all tools would have to be manually installed, often a source of great frustration and a key step where reproducibility between analyses is lost.
+All nf-core pipelines fully utilise the built-in support for software packaging that Nextflow offers - namely Docker, Singularity and Conda:
````
This isnt entirely true. Some pipelines like `imcyto` dont offer all of these options because it impossible to create a Conda environment with all of the pipeline dependencies.

drpatelh commented on markdown/usage/configuration.md:
```diff
@@ -0,0 +1,43 @@
+It has built-in support for HPC execution schedulers such as Slurm, SGE, PBS, LSF and more as well as cloud compute infrastructure such as AWS Batch and Google Cloud.
+
+In order to get nf-core pipelines to run properly on your system, you will need to configure Nextflow so that it knows how best to run your analysis jobs.
+Each nf-core pipeline comes with "sensible defaults" for every configuration option. These are then overwritten as needed.
```
May need to mention something about `executor` here or later i.e. if you have a cluster available you can create a config file that defines the relevant `executor`

apeltzer replied:
I agree - maybe also link to the Nextflow Docs on Executors from here then?

ewels replied:
Sounds good. We had quite a bit of detail on executors and I removed it. But would be good to mention it and link to the NF docs, agreed 👍

drpatelh commented on markdown/usage/installation.md:
```diff
@@ -1,28 +1,15 @@
-Nextflow runs on most POSIX systems (Linux, Mac OSX etc). It can be installed by running the following commands:
+## Nextflow
+
+All nf-core pipelines use Nextflow - this is the only absolute requirement for running pipelines.
```
Unless you are running in `local` mode you would need to have `Docker`, `Singularity` or `Conda` installed too.

ewels replied:
Yes, but you _could_ have all of that software installed, so Nextflow is the only _absolute_ requirement 😉 

Ok, this is a bit unhelpful..

drpatelh commented on markdown/usage/installation.md:
```diff
@@ -31,125 +18,52 @@ java -version
+  * Does not need root access or any daemon processes - images built from files
+* [Conda](https://conda.io/)
+  * Packaging system that manages environments instead of running analysis in containers.
+  * The software still runs in your operating system environment, theoretically the reproducibility of results is slightly lower than Docker / Singularity due to this.
```
Also because the low-level dependencies required to create the environment can change over time.

drpatelh commented on markdown/usage/installation.md:
```diff
@@ -31,125 +18,52 @@ java -version
 ## Reference genomes
 
-See [Reference genomes](reference_genomes)
+Some pipelines come with built-in support for iGenomes references. It may be preferable for you to download a local copy of these to your system. For more information, see [_Reference genomes_](reference_genomes).
```
Maybe mention its better to download locally to overcome download issues is using the same reference  and to save storage space.

drpatelh commented on markdown/usage/introduction.md:
```diff
@@ -24,86 +24,52 @@ The benefits of having your pipeline in Nextflow include:
+## How to run a pipeline
 
-[Nextflow has built-in support for Conda](https://www.nextflow.io/docs/latest/conda.html) that allows us to manage dependencies using Conda recipes and environment files. The nf-core pipelines make extensive use of the bioinformatics conda channel [Bioconda](https://bioconda.github.io/) making it simple to install all required tools.
+Nextflow works best when you have an active internet connection, as it is able to fetch all pipeline reqiurements. If you need to run offline, please see [_Running offline_](offline.md).
```
```suggestion
Nextflow works best when you have an active internet connection, as it is able to fetch all pipeline requirements. If you need to run offline, please see [_Running offline_](offline.md).
```

drpatelh commented on markdown/usage/introduction.md:
```diff
@@ -24,86 +24,52 @@ The benefits of having your pipeline in Nextflow include:
-* _recommended:_ Docker engine 1.10.x (or higher)
+3. Configure Nextflow to run on your system.
 
+    * The simplest way to run is with `-profile docker` (or `singularity` /  `conda`) which will tell nextflow will execute jobs locally using Docker for software requirements.
```
```suggestion
    * The simplest way to run is with `-profile docker` (or `singularity` /  `conda`) which will tell Nextflow to execute jobs locally using Docker to fulfil the software requirements.
```

drpatelh commented on markdown/usage/introduction.md:
```diff
@@ -24,86 +24,52 @@ The benefits of having your pipeline in Nextflow include:
 
+    * The simplest way to run is with `-profile docker` (or `singularity` /  `conda`) which will tell nextflow will execute jobs locally using Docker for software requirements.
 
+    * For more complex configuration of nextflow for your system, please see the [_Nextflow configuration_](https://nf-co.re/usage/configuration) documentation.
```
```suggestion
    * For more complex configuration of Nextflow for your system, please see the [_Nextflow configuration_](https://nf-co.re/usage/configuration) documentation.
```

drpatelh commented on markdown/usage/introduction.md:
```diff
@@ -24,86 +24,52 @@ The benefits of having your pipeline in Nextflow include:
+    * For more complex configuration of nextflow for your system, please see the [_Nextflow configuration_](https://nf-co.re/usage/configuration) documentation.
 
-1. Install Nextflow
+4. To test that everything is working properly, try running the tests for you your pipeline of interest in your terminal:
```
```suggestion
4. To test that everything is working properly, try running the tests for your pipeline of interest in the terminal:
```

drpatelh commented on markdown/usage/introduction.md:
````diff
@@ -24,86 +24,52 @@ The benefits of having your pipeline in Nextflow include:
     ```
 
-2. Install nf-core tools
+    * Replace `<pipeline_name>` with the name of a nf-core pipeline.
````
```suggestion
    * Replace `<pipeline_name>` with the name of an nf-core pipeline.
```

drpatelh commented on markdown/usage/introduction.md:
````diff
@@ -24,86 +24,52 @@ The benefits of having your pipeline in Nextflow include:
-    ```bash
-    nf-core list
-    ```
+    * If you don't have Docker installed, replace `docker` in the command with either `singularity` or `conda`.
````
Move this point up to 2nd in the listing.

drpatelh commented on markdown/usage/introduction.md:
```diff
@@ -24,86 +24,52 @@ The benefits of having your pipeline in Nextflow include:
+    * If you don't have Docker installed, replace `docker` in the command with either `singularity` or `conda`.
 
-4. To test that everything required is installed, try running a pipeline test in your terminal
+    * If the pipeline fails, check the [troubleshooting docs](/usage/troubleshooting.md) and ask for help on the nf-core Slack (see [https://nf-co.re/join](https://nf-co.re/join)).
```
```suggestion
    * If the pipeline fails, check the [troubleshooting docs](/usage/troubleshooting.md) and ask for help on the nf-core Slack channel for that particular pipeline (see [https://nf-co.re/join](https://nf-co.re/join)).
```

drpatelh commented on markdown/usage/introduction.md:
```diff
@@ -24,86 +24,52 @@ The benefits of having your pipeline in Nextflow include:
+## Tips and tricks
 
-    You will find the specific parameters required for each pipeline in the documentation of the respective pipeline.
+* Hyphens matter! Core nextflow options use one (`-`), pipeline specific paramters use two (`--`)
```
```suggestion
* Hyphens matter! Core Nextflow command-line options use one (`-`) whereas pipeline specific parameters use two (`--`)
```

drpatelh commented on markdown/usage/introduction.md:
```diff
@@ -24,86 +24,52 @@ The benefits of having your pipeline in Nextflow include:
+* Hyphens matter! Core nextflow options use one (`-`), pipeline specific paramters use two (`--`)
+* Specify `--email your@email.com` to receive emails when your pipeline completes
+* Always specify `-r <version-number>` when running to explicitly use a specific release. Then an identical command can be used in the future to give identical results.
+* Use `-resume` to restart pipelines that did not complete: successful tasks skipped using the outputs from the previous run.
```
```suggestion
* Use `-resume` to restart pipelines that did not complete. This ensures that successful tasks from the previous run wont be re-executed.
```

drpatelh commented on markdown/usage/offline.md:
```diff
@@ -0,0 +1,44 @@
+To help with this process, we have created a helper tool as part of the _nf-core_ package to automate this for you.
+
+On a computer with an internet connection, run `nf-core download <pipeline>` to download the pipeline and config profiles.
+Add the argument `--singularty` to also fetch the singularity container (requires Singularity to be installed on your system).
```
```suggestion
Add the argument `--singularity` to also fetch the singularity container (requires Singularity to be installed on your system).
```

ewels replied:
Hawk eyes!! 👀

drpatelh commented on markdown/usage/reference_genomes.md:
```diff
@@ -3,18 +3,34 @@ title: Reference Genomes
-Read [Adding your own cluster configuration](adding_own_config) to find out how to set up custom config files.
+## Illumina iGenomes
+To make the use of reference genomes easier, Illumina has developed a centralised resource called [iGenomes](https://support.illumina.com/sequencing/sequencing_software/igenome.html).
+Multiple reference index types are held together with consistent structure for multiple genomes.
```
```suggestion
The most commonly used reference files are organised in a consistent structure for multiple genomes.
```

drpatelh commented on markdown/usage/reference_genomes.md:
```diff
@@ -3,18 +3,34 @@ title: Reference Genomes
+To make the use of reference genomes easier, Illumina has developed a centralised resource called [iGenomes](https://support.illumina.com/sequencing/sequencing_software/igenome.html).
+Multiple reference index types are held together with consistent structure for multiple genomes.
+
+We have put a copy of iGenomes up onto AWS S3 hosting and this pipeline is configured to use this by default.
```
```suggestion
We have uploaded a copy of iGenomes onto AWS S3 and nf-core pipelines are configured to use this by default.
```

ewels replied:
![copypasta](https://media1.giphy.com/media/WI46AF0uiswes/giphy.gif)

drpatelh commented on markdown/usage/reference_genomes.md:
```diff
@@ -3,18 +3,34 @@ title: Reference Genomes
-To make this easier, the pipeline comes configured to understand reference genome keywords which correspond to preconfigured paths, meaning that you can just specify `--genome ID` when running the pipeline.
 
-Note that this genome key can also be specified in a config file if you always use the same genome.
+If you can not use the AWS-iGenomes references, you are always able to supply reference genome paths on the command line at run time.
```
```suggestion
If you are unable to use the AWS-iGenomes references, you can still supply reference genome paths on the command line via the pipeline's parameters e.g. `--fasta` or `--gtf`.
```

drpatelh commented:
I think we should remove the Nextflow documentation because it will become outdated and possibly point people to a more thorough resource. Could still keep the page but have a sneaky message saying this is where you should look?

apeltzer commented:
> I think we should remove the Nextflow documentation because it will become outdated and possibly point people to a more thorough resource. Could still keep the page but have a sneaky message saying this is where you should look?

Yeah, I think so too 👍

ewels commented:
Thanks @apeltzer and @drpatelh - all points now addressed I believe 👍 

I replaced the Nextflow tutorial with a page containing useful Nextflow resources - hopefully this will be helpful for people coming from Google. The page is not linked to from any navigation though, so over time I imagine that it will slowly die and we can later delete it if we want.

MaxUlysse commented on markdown/usage/configuration.md:
```diff
@@ -0,0 +1,47 @@
+2. [Shared _nf-core/configs_ configuration profiles](#shared-nf-core-configs)
+3. [Custom configuration files](#custom-configuration-files)
+
+Note that most users will need to specify an "executor" - telling Nextflow how to subject jobs to a job scheduler (_e.g._ SGE, LSF, SLURM, PBS, AWS Batch _etc._).
```
```suggestion
Note that most users will need to specify an "executor" - telling Nextflow how to submit jobs to a job scheduler (_e.g._ SGE, LSF, SLURM, PBS, AWS Batch _etc._).
```
I prefer submit over subject, but that's personnal, not sure what is the best way.
Feel free to dismiss

ewels replied:
Subject is a typo 😓 Submit is definitely the correct word 👍

MaxUlysse commented on markdown/usage/configuration.md:
```diff
@@ -0,0 +1,47 @@
+## Custom configuration files
+
+If you are the only person to be running this pipeline, you can create a local config file and use this.
+Nextflow looks for these files in three locations (in this order):
```
And which one takes priority?
The first one, or the later?

ewels replied:
"in this order" - does this not suggest first to last?

MaxUlysse replied:
Given the same order, the first file could have priority, or the last one could take it.
I find it confusing not to explicitly specify which one could take priority

ewels replied:
Added a new paragraph under the list with examples to make this as clear as possible 👍

MaxUlysse commented on markdown/usage/configuration.md:
```diff
@@ -0,0 +1,47 @@
+
+1. User's home directory: `~/.nextflow/config`
+2. Analysis working directory: `nextflow.config`
+3. Custom path specified on the command line: `-c path/to/config` (multiple can be given)
```
What is the priority order for multiple files?

ewels replied:
The order given..

ewels replied:
See new examples underneath 👍

MaxUlysse commented on markdown/usage/installation.md:
```diff
@@ -31,125 +17,60 @@ java -version
 
-This uses a number of sensible defaults for process requirements and is suitable for running
-on a simple (if powerful!) local server.
+Updating nextflow is as simple as running `nextflow self-update`.
```
If you specify how to install with condat, should we specify how to do a Nextflow update with conda as well?

ewels replied:
Yeah, I guess so. Does `nextflow self-update` not work with conda? Probably not I guess..

MaxUlysse replied:
Should we just drop the conda install command and link the bioconda nextflow page?

ewels replied:
I started doing this but then we still have the problem with the self-update command. Instead I've just added mention of `conda update nextflow`..

MaxUlysse commented on markdown/usage/installation.md:
```diff
@@ -31,125 +17,60 @@ java -version
+Updating nextflow is as simple as running `nextflow self-update`.
 
-Be warned of two important points about this default configuration:
+Once installed you will probably need to configure Nextflow to run on your system. For instructions, see [_Nextflow configuration_](configuration).
```
Is this a link to something?
Did you forgot a `#`?

ewels replied:
It is a link, to a page called `configuration`. Should work I think.. Can make it `configuration.md` if you prefer so that it works on both GitHub and the website. We are not very consistent in this though..

MaxUlysse replied:
I see, can we do `configuration.md` then, if I remember well you remove `.md` from the pages and links?

MaxUlysse replied:
cf https://github.com/nf-core/nf-co.re/pull/259/files#diff-f18d7381b1d628ed6e8a7a66cb1dbb8cR29

ewels replied:
The `.md` isn't (currently) removed from the URL, but the page does load. For example: https://nf-co.re/usage/installation.md

I can probably add a rule to redirect if .md is specified though, so that the URL is kept clean in the browser.

ewels replied:
✅ Added new htaccess rules to do automatic redirection
✅ Uses GitHub friendly markdown filenames everywhere now, I think..

MaxUlysse commented on markdown/usage/installation.md:
```diff
@@ -31,125 +17,60 @@ java -version
+  * Typically used locally / on single-user servers and the cloud.
+  * Analysis runs in a _container_, which behaves like an isolated operating system
+  * Previously required system root access, though a "rootless mode" is available since late 2019
+  * Best option for ultimate reproducibility
```
How does it compare with singularity?
You're saying here that Docker is the best for that, and in the Conda part it seems that Docker and Singularity are at the same level

drpatelh replied:
Yeah. I agree. Is it any better than `Singularity` for reproducibility?

ewels replied:
Probably not..

ewels replied:
Removed this, and made the conda lower reproducibility argument a bit stronger / clearer.

MaxUlysse commented on markdown/usage/introduction.md:
````diff
@@ -24,86 +24,58 @@ The benefits of having your pipeline in Nextflow include:
-    ```bash
-    nf-core list
-    ```
+    * If the pipeline fails, check the [troubleshooting docs](/usage/troubleshooting.md) and ask for help on the nf-core Slack channel for that particular pipeline (see [https://nf-co.re/join](https://nf-co.re/join)).
````
```suggestion
    * If the pipeline fails, check the [troubleshooting docs](troubleshooting.md) and ask for help on the nf-core Slack channel for that particular pipeline (see [https://nf-co.re/join](https://nf-co.re/join)).
```
I think we're already in `usage/` here in `introduction.md`, so the link should be `troubleshooting.md`

ewels replied:
True. Should work on the website, as the first `/` sets the path to the web root. But it'll probably break on GitHub.

MaxUlysse commented on markdown/usage/introduction.md:
````diff
@@ -24,86 +24,58 @@ The benefits of having your pipeline in Nextflow include:
 
     ```bash
-    nextflow run nf-core/<pipeline_name> -profile standard,docker [parameters]
+    nextflow run nf-core/methylseq -profile docker -reads 'input_data/*.fastq.gz' --outdir myproj/results --genome GRCh38
````
```suggestion
    nextflow run nf-core/methylseq -profile docker --reads 'input_data/*.fastq.gz' --outdir myproj/results --genome GRCh38
```

MaxUlysse commented on markdown/usage/offline.md:
```diff
@@ -0,0 +1,44 @@
+
+Some pipelines require reference genomes and have built-in integration of AWS-iGenomes.
+If you wish to use these references you must download them and transfer to your offline cluster.
+Once transferred, follow the [reference genomes documentation](reference_genomes) to configure the base path for the references.
```
```suggestion
Once transferred, follow the [reference genomes documentation](reference_genomes.md) to configure the base path for the references.
```

MaxUlysse commented on markdown/usage/reference_genomes.md:
```diff
@@ -34,24 +50,6 @@ params {
-## Help
-
-In any case if you need help with this, please don't hesitate to ask in our dedicated [Slack](https://nf-co.re/join/slack) channels: [`#igenomes`](https://nfcore.slack.com/channels/igenomes) or [`#help`](https://nfcore.slack.com/channels/help).
+Read the [Nextflow configuration documentation](/usage/configuration) for more information about custom config files.
```
```suggestion
Read the [Nextflow configuration documentation](configuration.md) for more information about custom config files.
```

drpatelh commented:
Ok. Im happy when @MaxUlysse is :+1:

ewels commented:
Ok, all comments now addressed I believe 👍
